File: connectors/peertube-ingest/app/metadata.py
"""Extracts meeting metadata (title, date, participants) from the MinIO object name
(study 2.12 line 589: "meeting metadata association -- title, date,
participants -- extracted from the object name or S3 tags"). Pure functions,
testable without network access or an S3 SDK.

Expected naming convention, on the LiveKit Egress export side (to be documented/
configured at the Egress rule level):
  <ISO-date>_<title-slug>_<participant1>-<participant2>-....<ext>
e.g.: "2026-09-05_kickoff-projet-libre365_alice-bob-carol.mp4"

If the name does not follow this format, falls back to a title derived from the
raw file name, with no date or participants -- degrades gracefully rather than failing.
"""

from __future__ import annotations

import re
from typing import Dict

from .types import MeetingMetadata

_NAME_RE = re.compile(
    r"^(\d{4}-\d{2}-\d{2})_([a-z0-9-]+)_([a-z0-9-]+(?:-[a-z0-9-]+)*)$",
    re.IGNORECASE,
)


def extract_meeting_metadata_from_key(object_key: str) -> MeetingMetadata:
    file_name = object_key.split("/")[-1] if object_key else object_key
    without_ext = re.sub(r"\.[a-zA-Z0-9]+$", "", file_name)

    match = _NAME_RE.match(without_ext)
    if not match:
        title = re.sub(r"[-_]", " ", without_ext).strip() or "Meeting recording"
        return MeetingMetadata(title=title, date=None, participants=[])

    iso_date, title_slug, participants_slug = match.groups()
    return MeetingMetadata(
        title=_slug_to_title(title_slug),
        date=iso_date,
        participants=[_slug_to_title(p) for p in participants_slug.split("-")],
    )


def merge_with_s3_tags(base: MeetingMetadata, tags: Dict[str, str]) -> MeetingMetadata:
    """Additionally applies S3 tags (`meeting-title`, `meeting-participants`) when
    present, taking priority over what is inferred from the file name."""
    participants = (
        [p.strip() for p in tags["meeting-participants"].split(",")]
        if "meeting-participants" in tags
        else base.participants
    )
    return MeetingMetadata(
        title=tags.get("meeting-title", base.title),
        date=tags.get("meeting-date", base.date),
        participants=participants,
    )


def _slug_to_title(slug: str) -> str:
    return " ".join(word[:1].upper() + word[1:] for word in slug.split("-") if word)

File: connectors/peertube-ingest/src/metadata.ts
import { MeetingMetadata } from "./types";

/**
 * Extrait les metadonnees de reunion (titre, date, participants) depuis le nom d'objet
 * MinIO (etude 2.12 ligne 589 : "association des metadonnees de reunion -- titre, date,
 * participants -- extraites du nom d'objet ou de tags S3"). Fonction pure, testable
 * sans reseau ni SDK S3.
 *
 * Convention de nommage attendue, cote export LiveKit Egress (a documenter/configurer
 * au niveau de la regle d'Egress) :
 *   <ISO-date>_<slug-titre>_<participant1>-<participant2>-....<ext>
 * ex: "2026-09-05_kickoff-projet-open365_alice-bob-carol.mp4"
 *
 * Si le nom ne suit pas ce format, retombe sur un titre derive du nom de fichier brut,
 * sans date ni participants -- degrade proprement plutot que d'echouer.
 */
export function extractMeetingMetadataFromKey(objectKey: string): MeetingMetadata {
  const fileName = objectKey.split("/").pop() ?? objectKey;
  const withoutExt = fileName.replace(/\.[a-zA-Z0-9]+$/, "");

  const match = withoutExt.match(
    /^(\d{4}-\d{2}-\d{2})_([a-z0-9-]+)_([a-z0-9-]+(?:-[a-z0-9-]+)*)$/i
  );

  if (!match) {
    return {
      title: withoutExt.replace(/[-_]/g, " ").trim() || "Enregistrement de reunion",
      date: null,
      participants: [],
    };
  }

  const [, isoDate, titleSlug, participantsSlug] = match;
  return {
    title: slugToTitle(titleSlug),
    date: isoDate,
    participants: participantsSlug.split("-").map((p) => slugToTitle(p)),
  };
}

/** Applique en complement les tags S3 (`meeting-title`, `meeting-participants`) s'ils existent,
 * prioritaires sur ce qui est deduit du nom de fichier. */
export function mergeWithS3Tags(
  base: MeetingMetadata,
  tags: Record<string, string>
): MeetingMetadata {
  return {
    title: tags["meeting-title"] ?? base.title,
    date: tags["meeting-date"] ?? base.date,
    participants: tags["meeting-participants"]
      ? tags["meeting-participants"].split(",").map((p) => p.trim())
      : base.participants,
  };
}

function slugToTitle(slug: string): string {
  return slug
    .split("-")
    .filter(Boolean)
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
    .join(" ");
}
